tests/test_utils.py

```diff
@@ -63,6 +65,33 @@
 U_toffoli[6:8, 6:8] = np.array([[0, 1], [1, 0]])
 
 
+test_Hamiltonians = [np.array([[2.5, -0.5], [-0.5, 2.5]]), np.array(np.diag([0, 0, 0, 1]))]
+
+
+class TestDecomposition:
+    """Tests the decompose_hamiltonian function"""
+
+    @pytest.mark.parametrize("H", test_Hamiltonians)
+    def test_decomp(self, H):
+        """Tests that decompose_hamiltonian successfully decomposes Hamiltonians into Pauli matrices"""
+        N = int(np.log2(len(H)))
+        coeff, combs = pu.decompose_hamiltonian(H)
+        result = np.zeros((2 ** N, 2 ** N))
+        matrices = []
+        for term in combs:
+            base = [qml.Identity] * N
+            if term.num_wires == N:
+                matrices.append(term.matrix)
+            else:
+                base[term.wires[0]] = term.__class__
+                matrices.append(
+                    functools.reduce(operator.matmul, [t(i) for i, t in enumerate(base)]).matrix
+                )
+        for term in zip(coeff, matrices):
+            result = result + term[0] * term[1]
```

Comment:
This is quite neat; essentially you are decomposing a Hamiltonian using the function, and then _evaluating the linear combination_ by making use of the `observable.matrix` property. Nice! However, it could be done in a slightly cleaner manner, without re-using the logic from the function to be tested, like so:

```python
mats = []

for c, o in zip(coeff, obs_list):
    base = [np.identity(2)] * N

    if isinstance(o, qml.operation.Tensor):
        for obs in o.obs:
            base[obs.wires[0]] = obs.matrix
    else:
        base[o.wires[0]] = o.matrix

    mats.append(c*functools.reduce(np.kron, base))

result = np.linalg.multi_dot(mats)
```

_Note: I have note checked the above for correctness, it could need adjusting_